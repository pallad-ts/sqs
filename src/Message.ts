import * as AWS from 'aws-sdk';
import {TypedArray} from "./typedArray";
import {Queue} from "./Queue";
import {ulid} from "ulid";
import {isBinaryKind} from "./DataType";

export class Message<TBody = string, TAttributes = Message.Attributes> {
    constructor(public readonly raw: AWS.SQS.Message,
                public readonly queue: Queue.Info,
                public readonly body: TBody,
                public readonly attributes: TAttributes) {
        Object.freeze(this);
    }

    get groupId(): string | undefined {
        return (this.raw.Attributes as AWS.SQS.MessageSystemAttributeMap).MessageGroupId;
    }

    get deduplicationId(): string | undefined {
        return (this.raw.Attributes as AWS.SQS.MessageSystemAttributeMap).MessageDeduplicationId;
    }

    get internalAttributes() {
        return this.raw.Attributes;
    }

    get sequenceNumber(): string | undefined {
        if (this.raw.Attributes) {
            return this.raw.Attributes['SequenceNumber'] as string;
        }
    }

    toInput(delay?: number): Message.Input {
        let attributes: { [key: string]: Message.Input.MessageAttribute } | undefined = undefined;

        if (this.raw.MessageAttributes) {
            attributes = {};
            for (const [key, value] of Object.entries(this.raw.MessageAttributes)) {
                attributes[key] = {
                    type: value.DataType,
                    value: isBinaryKind(value.DataType) ? value.BinaryValue : value.StringValue
                };
            }
        }
        return {
            body: this.body,
            attributes: attributes,
            deduplicationId: this.deduplicationId,
            groupId: this.groupId,
            delay
        };
    }
}

export namespace Message {
    export type Attributes = { [key: string]: any };

    export interface Input<TBody = any> {
        delay?: number;
        body: TBody;
        attributes?: { [key: string]: Input.MessageAttributeValue };
        groupId?: string;
        deduplicationId?: string;
    }

    export namespace Input {
        export function toBatchInput(input: Input) {
            return {
                ...input,
                id: ulid()
            }
        }
    }

    export interface BatchInput<TBody = any> extends Input<TBody> {
        id: string;
    }

    export function isBatchInput(value: BatchInput | Input): value is BatchInput {
        return 'id' in value;
    }

    export namespace Input {
        export type MessageAttributeValue = string | number | Buffer | TypedArray | MessageAttribute;
        export type MessageAttribute = { type: string, value: any };
    }
}