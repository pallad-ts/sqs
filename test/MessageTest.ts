import * as fixtures from '../fixtures';
import {Message} from "@src/Message";
import {SQS} from "aws-sdk";

describe('Message', () => {
    const MESSAGE_STANDARD = fixtures.message.regular;
    const MESSAGE_STANDARD_NO_ATTRIBUTES = fixtures.message.noAttributes;
    const MESSAGE_FIFO = fixtures.message.fifo;
    const MESSAGE_FIFO_DEDUP = fixtures.message.fifoDedup;

    describe('structure', () => {

        it('standard', () => {
            const message = MESSAGE_STANDARD;
            expect(message.groupId)
                .toBeUndefined();

            expect(message.deduplicationId)
                .toBeUndefined();

            expect(message.internalAttributes)
                .toEqual(message.raw.Attributes);
        });

        it('fifo', () => {
            const message = MESSAGE_FIFO;
            expect(message.groupId)
                .toEqual(
                    (message.raw.Attributes as SQS.MessageSystemAttributeMap).MessageGroupId
                );

            expect(message.groupId)
                .toBeDefined();

            expect(message.deduplicationId)
                .toBeUndefined();

            expect(message.internalAttributes)
                .toEqual(message.raw.Attributes);
        });

        it('fifo dedup', () => {
            const message = MESSAGE_FIFO_DEDUP;
            expect(message.groupId)
                .toEqual(
                    (message.raw.Attributes as SQS.MessageSystemAttributeMap).MessageGroupId
                );

            expect(message.groupId)
                .toBeDefined();

            expect(message.deduplicationId)
                .toEqual(
                    (message.raw.Attributes as SQS.MessageSystemAttributeMap).MessageDeduplicationId
                );

            expect(message.deduplicationId)
                .toBeDefined();

            expect(message.internalAttributes)
                .toEqual(message.raw.Attributes);
        });
    });

    describe('converting to input', () => {
        describe.each<[string, Message<string, any>]>([
            ['standard', MESSAGE_STANDARD],
            ['standard without attributes', MESSAGE_STANDARD_NO_ATTRIBUTES],
            ['fifo', MESSAGE_FIFO],
            ['fifo with dedup', MESSAGE_FIFO_DEDUP]
        ])('message %s', (_s, message) => {
            it('with delay', () => {
                expect(message.toInput(100)).toMatchSnapshot()
            });

            it('without delay', () => {
                expect(message.toInput()).toMatchSnapshot();
            });
        });
    });

    describe('Input', () => {
        it('converting to batch input', () => {
            const input = MESSAGE_STANDARD.toInput();
            expect(Message.Input.toBatchInput(input))
                .toMatchObject({
                    ...input,
                    id: expect.stringMatching(/[A-Z0-9]{26}/)
                })
        });

        it('is batch input', () => {
            expect(Message.isBatchInput(MESSAGE_STANDARD.toInput()))
                .toBeFalsy();

            expect(Message.Input.toBatchInput(MESSAGE_STANDARD.toInput()))
                .toBeTruthy();
        });
    });
});