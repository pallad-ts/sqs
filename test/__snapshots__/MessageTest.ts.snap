// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message converting to input message fifo with dedup with delay 1`] = `
Object {
  "attributes": Object {
    "any": Object {
      "type": "Number",
      "value": "100",
    },
    "bin": Object {
      "type": "Binary",
      "value": Object {
        "data": Array [
          116,
          101,
          115,
          116,
        ],
        "type": "Buffer",
      },
    },
    "some": Object {
      "type": "String",
      "value": "string",
    },
  },
  "body": "some-body",
  "deduplicationId": "dedupid",
  "delay": 100,
  "groupId": "groupid",
}
`;

exports[`Message converting to input message fifo with dedup without delay 1`] = `
Object {
  "attributes": Object {
    "any": Object {
      "type": "Number",
      "value": "100",
    },
    "bin": Object {
      "type": "Binary",
      "value": Object {
        "data": Array [
          116,
          101,
          115,
          116,
        ],
        "type": "Buffer",
      },
    },
    "some": Object {
      "type": "String",
      "value": "string",
    },
  },
  "body": "some-body",
  "deduplicationId": "dedupid",
  "delay": undefined,
  "groupId": "groupid",
}
`;

exports[`Message converting to input message fifo with delay 1`] = `
Object {
  "attributes": Object {
    "any": Object {
      "type": "Number",
      "value": "100",
    },
    "bin": Object {
      "type": "Binary",
      "value": Object {
        "data": Array [
          116,
          101,
          115,
          116,
        ],
        "type": "Buffer",
      },
    },
    "some": Object {
      "type": "String",
      "value": "string",
    },
  },
  "body": "some-body",
  "deduplicationId": undefined,
  "delay": 100,
  "groupId": "groupid",
}
`;

exports[`Message converting to input message fifo without delay 1`] = `
Object {
  "attributes": Object {
    "any": Object {
      "type": "Number",
      "value": "100",
    },
    "bin": Object {
      "type": "Binary",
      "value": Object {
        "data": Array [
          116,
          101,
          115,
          116,
        ],
        "type": "Buffer",
      },
    },
    "some": Object {
      "type": "String",
      "value": "string",
    },
  },
  "body": "some-body",
  "deduplicationId": undefined,
  "delay": undefined,
  "groupId": "groupid",
}
`;

exports[`Message converting to input message standard with delay 1`] = `
Object {
  "attributes": Object {
    "any": Object {
      "type": "Number",
      "value": "100",
    },
    "bin": Object {
      "type": "Binary",
      "value": Object {
        "data": Array [
          116,
          101,
          115,
          116,
        ],
        "type": "Buffer",
      },
    },
    "some": Object {
      "type": "String",
      "value": "string",
    },
  },
  "body": "some-body",
  "deduplicationId": undefined,
  "delay": 100,
  "groupId": undefined,
}
`;

exports[`Message converting to input message standard without attributes with delay 1`] = `
Object {
  "attributes": undefined,
  "body": "some-body",
  "deduplicationId": undefined,
  "delay": 100,
  "groupId": undefined,
}
`;

exports[`Message converting to input message standard without attributes without delay 1`] = `
Object {
  "attributes": undefined,
  "body": "some-body",
  "deduplicationId": undefined,
  "delay": undefined,
  "groupId": undefined,
}
`;

exports[`Message converting to input message standard without delay 1`] = `
Object {
  "attributes": Object {
    "any": Object {
      "type": "Number",
      "value": "100",
    },
    "bin": Object {
      "type": "Binary",
      "value": Object {
        "data": Array [
          116,
          101,
          115,
          116,
        ],
        "type": "Buffer",
      },
    },
    "some": Object {
      "type": "String",
      "value": "string",
    },
  },
  "body": "some-body",
  "deduplicationId": undefined,
  "delay": undefined,
  "groupId": undefined,
}
`;
